### PR TITLE
SWC-6896 - Refactor AccessRequirementTable to use @tanstack/react-table

### PR DESCRIPTION
**Pull request by nickgros**
Addition of feature will be in a later PR

### CODE REVIEW

File: packages/synapse-react-client/src/components/Authentication/StandaloneLoginForm/Login.stories.ts [hunk 1]
```diff
@@ -21,5 +22,6 @@ export const LoginDemo: Story = {
         )
       })
     },
+    onPasswordLoginSelected: fn(),
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/Authentication/StandaloneLoginForm/Login.stories.ts:25]: Fix issue where adding this prop broke logging into Storybook

File: packages/synapse-react-client/src/components/SynapseTable/SynapseTable.integration.test.tsx [hunk 2]
```diff
@@ -412,9 +412,9 @@ describe('SynapseTable tests', () => {
     // simulate having clicked the sort button on the first column, projectName
     const sortedColumn = 'projectName'
 
-    const sortButton = (
-      await screen.findAllByRole('button', { name: 'sort' })
-    )[0]
+    const sortButton = await screen.findByRole('button', {
+      name: 'Sort by Project Name',
+    })
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/SynapseTable/SynapseTable.integration.test.tsx:417]: Changed shared component label

File: packages/synapse-react-client/src/components/dataaccess/AccessRequirementTable.integration.test.tsx [hunk 3]
```diff
@@ -313,13 +315,15 @@ describe('Access Requirement Table tests', () => {
...
+    // clicking a different column (name) should sort by that column, ascending
+    const nameSortButton = screen.getByLabelText(
+      'Sort by Access Requirement Name',
+    )
     await userEvent.click(nameSortButton)
     await waitFor(() =>
       expect(onServiceRecievedRequest).toHaveBeenLastCalledWith(
         expect.objectContaining({
-          sort: [{ field: 'NAME', direction: 'DESC' }],
+          sort: [{ field: 'NAME', direction: 'ASC' }],
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/dataaccess/AccessRequirementTable.integration.test.tsx:326]: @tanstack/react-table is kinda smart, it makes the default initial sort for strings ascending, which is more intuitive for the user. Update the test accordingly.

File: packages/synapse-react-client/src/components/dataaccess/AccessRequirementTable.tsx [hunk 4]
```diff
@@ -70,14 +163,36 @@ export function AccessRequirementTable(props: AccessRequirementTableProps) {
...
 
-  const [sort, setSort] = useState<AccessRequirementSearchSort>({
-    field: 'CREATED_ON',
-    direction: 'DESC',
-  })
+  const [sortingState, setSortingState] = useState<SortingState>(
+    DEFAULT_SORTING_STATE,
+  )
+
+  const requestSortState: AccessRequirementSearchRequest['sort'] =
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/dataaccess/AccessRequirementTable.tsx:170]: map the @tanstack/react-table `SortingState` into the correct API object

File: packages/synapse-react-client/src/components/dataaccess/AccessRequirementTable.tsx [hunk 5]
```diff
@@ -94,24 +209,37 @@ export function AccessRequirementTable(props: AccessRequirementTableProps) {
...
+      columns: columns,
+      state: {
+        sorting: sortingState,
+      },
+      onSortingChange: setSortingState,
+      getRowId: row => row.id,
+      getCoreRowModel: getCoreRowModel(),
+      columnResizeMode: 'onChange',
+      manualSorting: true,
+      enableMultiSort: false,
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/dataaccess/AccessRequirementTable.tsx:236]: We could easily toggle multi-sort on, but for now it's not very useful

File: packages/synapse-react-client/src/components/dataaccess/AccessRequirementTable.tsx [hunk 6]
```diff
@@ -128,86 +256,71 @@ export function AccessRequirementTable(props: AccessRequirementTableProps) {
...
-                  </td>
-                  <td>
-                    {ar.relatedProjectIds.map(projectId => (
-                      <React.Fragment key={projectId}>
-                        <EntityLink entity={projectId} />{' '}
-                        <span className="InlineLabel">({projectId})</span>
-                        <br />
-                      </React.Fragment>
+      <div>
+        <StyledTableContainer
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/dataaccess/AccessRequirementTable.tsx:260]: Use shared styled component

File: packages/synapse-react-client/src/components/dataaccess/AccessRequirementTable.tsx [hunk 7]
```diff
@@ -94,24 +209,37 @@ export function AccessRequirementTable(props: AccessRequirementTableProps) {
...
+      getCoreRowModel: getCoreRowModel(),
+      columnResizeMode: 'onChange',
+      manualSorting: true,
+      enableMultiSort: false,
+      // Sort is required for this API call, so prevent the user from removing the sort.
+      enableSortingRemoval: false,
+    })
 
   return (
-    <div className="bootstrap-4-backport">
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/dataaccess/AccessRequirementTable.tsx:114]: Goodbye bootstrap

File: packages/synapse-react-client/src/components/styled/ColumnHeader.tsx [hunk 8]
```diff
@@ -61,11 +61,11 @@ export default function ColumnHeader<TData = unknown, TValue = unknown>(
         )}
         {column.getCanFilter() && filterControl}
         {column.getCanSort() && (
-          <Tooltip title={`Sort ${title}`} placement={'top'}>
+          <Tooltip title={`Sort by ${title}`} placement={'top'}>
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/styled/ColumnHeader.tsx:64]: Improve the title of the button

File: packages/synapse-react-client/src/style/components/_governance-dashboard.scss [hunk 9]
```diff
@@ -30,8 +30,7 @@
 }
 
 .AccessApprovalsTable,
-.AccessSubmissionTable,
-.AccessRequirementsTable {
```
Comment (nickgros) [inline packages/synapse-react-client/src/style/components/_governance-dashboard.scss:34]: CSS class for this table is no longer necessary

File: packages/synapse-react-client/src/components/Forum/ForumTable.tsx [hunk 10]
```diff
@@ -153,70 +152,10 @@ export const ForumTable: React.FC<ForumTableProps> = ({
...
-                })}
-              </tr>
-            ))}
-          </tbody>
-        </table>
-      </StyledTableContainer>
+      <StyledTanstackTable
+        table={table}
+        styledTableContainerProps={{ sx: { my: 2 } }}
+      />
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/Forum/ForumTable.tsx:158]: Refactored to create a single styled component that renders a TanStack table

File: packages/synapse-react-client/src/components/dataaccess/UseAccessRequirementTable.tsx [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+        sorting: sortingState,
+      },
+      onSortingChange: setSortingState,
+      getRowId: row => row.id,
+      getCoreRowModel: getCoreRowModel(),
+      columnResizeMode: 'onChange',
+      manualSorting: true,
+      enableMultiSort: false,
+      // Sort is required for this API call, so prevent the user from removing the sort.
+      enableSortingRemoval: false,
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/dataaccess/UseAccessRequirementTable.tsx:232]: I initially tried to solve this manually but it turns out that seemingly every edge case is handled by this library!

File: packages/synapse-react-client/src/components/dataaccess/UseAccessRequirementTable.tsx [hunk 12]
```diff
@@ -0,0 +1,236 @@
...
+  },
+]
+type UseAccessRequirementTableOptions = {
+  nameOrID?: string
+  relatedProjectId?: string
+  reviewerId?: string
+  accessType?: ACCESS_TYPE
+}
+
+export function useAccessRequirementTable(
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/dataaccess/UseAccessRequirementTable.tsx:149]: Contains all logic for building the table & cell/header renderers

File: packages/synapse-react-client/src/components/dataaccess/UseAccessRequirementTable.tsx [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+      columns: columns,
+      state: {
+        sorting: sortingState,
+      },
+      onSortingChange: setSortingState,
+      getRowId: row => row.id,
+      getCoreRowModel: getCoreRowModel(),
+      columnResizeMode: 'onChange',
+      manualSorting: true,
+      enableMultiSort: false,
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/dataaccess/UseAccessRequirementTable.tsx:230]: We could easily toggle multi-sort on, but for now it's not very useful (the two existing sort fields, name and createdOn, never have duplicates) and it just makes it harder to reset your sort state.

File: packages/synapse-react-client/src/components/styled/StyledTableContainer.tsx [hunk 14]
```diff
@@ -1,8 +1,12 @@
 import { Box, BoxProps, styled } from '@mui/material'
 import { StyledComponent } from '@emotion/styled'
 
-export const StyledTableContainer: StyledComponent<BoxProps> = styled(Box)(
-  ({ theme }) => ({
+export type StyledTableContainerProps = BoxProps & {
+  density?: 'default' | 'compact'
+}
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/styled/StyledTableContainer.tsx:6]: Added `density` prop to make this more easily reusable and remove some common style overrides

File: packages/synapse-react-client/src/components/styled/StyledTableContainer.tsx [hunk 15]
```diff
@@ -19,5 +23,12 @@ export const StyledTableContainer: StyledComponent<BoxProps> = styled(Box)(
...
-)
+    ['th,td']: {
+      paddingLeft: theme.spacing(density === 'compact' ? 0 : 1),
+      paddingRight: theme.spacing(density === 'compact' ? 0 : 1),
+    },
+    td: {
+      paddingTop: theme.spacing(density === 'compact' ? 0 : 1),
+      paddingBottom: theme.spacing(density === 'compact' ? 0 : 1),
+    },
+  }))
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/styled/StyledTableContainer.tsx:34]: Add spacing to table cells by default

File: packages/synapse-react-client/src/components/dataaccess/AccessRequirementTable.tsx [hunk 16]
```diff
@@ -70,48 +23,16 @@ export function AccessRequirementTable(props: AccessRequirementTableProps) {
...
-      setSort({ field, direction: 'DESC' })
-    }
-  }
+  const { table, isLoading, hasNextPage, fetchNextPage } =
+    useAccessRequirementTable({
+      nameOrID,
+      relatedProjectId,
+      reviewerId,
+      accessType,
+    })
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/dataaccess/AccessRequirementTable.tsx:32]: data & table comes from this hook

File: packages/synapse-react-client/src/components/Forum/ForumTable.tsx [hunk 17]
```diff
@@ -153,70 +152,10 @@ export const ForumTable: React.FC<ForumTableProps> = ({
...
-                      )}
-                    </td>
-                  )
-                })}
-              </tr>
-            ))}
-          </tbody>
-        </table>
-      </StyledTableContainer>
+      <StyledTanStackTable
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/Forum/ForumTable.tsx:155]: Refactored to create a single styled component that renders a TanStack table

File: packages/synapse-react-client/src/components/EntityHeaderTable/EntityHeaderTable.tsx [hunk 18]
```diff
@@ -367,6 +367,7 @@ export const EntityHeaderTable = (props: EntityHeaderTableProps) => {
           sx={{
             th: {
               zIndex: 100,
+              maxHeight: '250px',
```
Comment (nickgros) [inline packages/synapse-react-client/src/components/EntityHeaderTable/EntityHeaderTable.tsx:370]: Pulled out of shared style and re-applied selectively where appropriate

**Review by jay-hodgson — APPROVED** [review]

**Pull request merged**